packages/firestore/src/local/indexeddb_index_manager.ts

```diff
@@ -516,6 +833,126 @@ export class IndexedDbIndexManager implements IndexManager {
       )
       .next(() => nextSequenceNumber);
   }
+
+  /**
+   * Applies notIn and != filters by taking the provided ranges and excluding
+   * any values that match the `notInValue` from these ranges. As an example,
+   * '[foo > 2 && foo != 3]` becomes  `[foo > 2 && < 3, foo > 3]`.
+   */
+  private applyNotIn(
+    indexRanges: IDBKeyRange[],
+    notInValues: Uint8Array[] | null
+  ): IDBKeyRange[] {
+    if (!notInValues) {
+      return indexRanges;
+    }
+
+    // The values need to be sorted so that we can return a sorted set of
+    // non-overlapping ranges.
+    notInValues.sort((l, r) => compareByteArrays(l, r));
+
+    const notInRanges: IDBKeyRange[] = [];
+    for (const indexRange of indexRanges) {
+      // Remove notIn values that are not applicable to this index range
+      const filteredRanges = notInValues.filter(
+        v =>
+          compareByteArrays(v, new Uint8Array(indexRange.lower[3])) >= 0 &&
+          compareByteArrays(v, new Uint8Array(indexRange.upper[3])) <= 0
+      );
+
+      if (filteredRanges.length === 0) {
+        notInRanges.push(indexRange);
+      } else {
+        // Use the existing bounds and interleave the notIn values. This means
+        // that we would split an existing range into multiple ranges that exclude
+        // the values from any notIn filter.
+        notInRanges.push(...this.interleaveRanges(indexRange, filteredRanges));
+      }
+    }
+    return notInRanges;
+  }
+
+  /**
+   * Splits up the range defined by `indexRange` and removes any values
+   * contained in `barrier`. As an example, if the original range is [1,4] and
+   * the barrier is [2,3], then this method would return [1,2), (2,3), (3,4].
+   */
+  private interleaveRanges(
+    indexRange: IDBKeyRange,
+    barriers: Uint8Array[]
+  ): IDBKeyRange[] {
+    const ranges: IDBKeyRange[] = [];
+
+    // The first index range starts with the lower bound and ends before the
+    // first barrier.
+    ranges.push(
+      IDBKeyRange.bound(
+        indexRange.lower,
+        this.generateNotInBound(indexRange.lower, barriers[0]),
```

Comment:
What if `barriers[0]` is larger than `indexRange.upper`, is that not possible?

I think we could do something like:

```
// Assume barriers is sorted, it is right?
let runningLower = indexRange.lower
let runningOpen = indexRange.lowerOpen
for(const barrier: barriers) {
  if(barrier <= indexRange.lower) {continue;}
  if(barrier >= indexRange.upper) {break;}
  ranges.push([runningLower /*need to handle lowerOpen as well*/, barrier]);
  runningLower = barrier
  lowerOpen = true
}

// Handles last
ranges.push([runningLower, indexRange.upper])
```

You could even do handle this by taking `IDBKeyRange[]`, sort both `indexRanges` and `barriers` together..then process from left to right. But maybe let's not go there, it should be rare in real world scenarios.

Reply:
I updated this code to work on browsers and Node. The browsers seem to do more validations, so I had to change some of the logic here. I also added a bunch of test code. Note that I myself don't fully grasp what is going on here anymore, but the tests are all passing and I think that I covered all corner cases.

Reply:
Still broken